README.md

Serialize to and deserialize from JSON in runtime. 
This pure Dart library does not depend on build time code generation.

## Installation
Add this dependency your pubspec.yaml file:

```
dependencies:
  enough_serialization: ^1.4.0
```
The latest version or `enough_serialization` is [![enough_serialization version](https://img.shields.io/pub/v/enough_serialization.svg)](https://pub.dartlang.org/packages/enough_serialization).


## Usage

You can choose between two serialization modes:

1. Extend `SerializableObject` or implement `Serializable` for full control and complex cases. You have to store your field values in a dynamic map, however.
2. Implement `OnDemandSerializable` to only write and read your field values to/from a dynamic map when needed. This comes with limitations when having deep nested structures - in that case any serialization names of complex fields must be unique.  

### Serialization with `Serializable` and `SerializableObject`

The easiest way is 
 * extend `SerializableObject`
 * define a `get` and `set` field for each supported attribute and retrieve value from / store values to the `attributes` map.
 * if you have nested objects such as lists, maps or other serializable objects, define a function that creates new object instances for the corresponding attribute name in the `objectCreators` field.
 * if you want to store other values such as an `enum`, register a transformation function in the `transformer` field. This can also be done for map keys and for map/list values.

#### Simple Example
When you have only basic fields such as `String`, `int`, `double`, `bool` the usage is straight forward:

```dart
import 'package:enough_serialization/enough_serialization.dart';

class SimpleArticle extends SerializableObject {
  String get name => attributes['name'];
  set name(String value) => attributes['name'] = value;

  int get price => attributes['price'];
  set price(int value) => attributes['price'] = value;

  double get popularity => attributes['popularity'];
  set popularity(double value) => attributes['popularity'] = value;
}

void main() {
  final article = SimpleArticle()
    ..name = 'Remote Control'
    ..price = 2499
    ..popularity = 0.1;
  final serializer = Serializer();
  final json = serializer.serialize(article);
  print('serialized article: $json');

  final inputJson =
      '{"name": "Remote Control", "price": 2499, "popularity": 0.1}';
  final deserializedArticle = SimpleArticle();
  serializer.deserialize(inputJson, deserializedArticle);
  print('name: ${deserializedArticle.name}');
  print('price: ${deserializedArticle.price}');
  print('popularity: ${deserializedArticle.popularity}');
}
```

#### Enum Example
Enumerations cannot be serialized or deserialized automatically. To support `enum` fields, register a `transformer` function:
```dart
transformers['area'] = (value) =>
        value is ArticleArea ? value.index : ArticleArea.values[value];
```
Here's a complete example:

```dart
enum ArticleArea { electronics, music }

class ArticleWithEnum extends SerializableObject {
  ArticleWithEnum() {
    transformers['area'] = (value) =>
        value is ArticleArea ? value.index : ArticleArea.values[value];
  }

  ArticleArea get area => attributes['area'];
  set area(ArticleArea value) => attributes['area'] = value;

  String get name => attributes['name'];
  set name(String value) => attributes['name'] = value;
}

void main() {
  final article = ArticleWithEnum()
    ..area = ArticleArea.electronics
    ..name = 'Remote Control';
  final serializer = Serializer();
  final json = serializer.serialize(article);
  print('serialized article: $json');

  final inputJson = '{"area": 0, "name": "Remote Control"}';
  final deserializedArticle = ArticleWithEnum();
  serializer.deserialize(inputJson, deserializedArticle);
  print('area: ${deserializedArticle.area}');
  print('name: ${deserializedArticle.name}');
}
```

#### Lists and Nested Serializable Objects
When you have nested objects, register a creation function in the `objectCreators` field. For nested maps and objects this function 
receives a `Map<String,dynamic>` parameter with the values of the nested child object. If needed, you can evaluate 
these values to determine which kind of object you need to create:

```dart
// create a list:
objectCreators['articles'] = (map) => <Article>[];
// create a nested simple object:
objectCreators['band'] = (map) => Band();
// created a nested complex field in a list:
objectCreators['articles.value'] = (map) {
      final int areaIndex = map['area'];
      final area = ArticleArea.values[areaIndex];
      switch (area) {
        case ArticleArea.electronics:
          return ElectronicsArticle();
        case ArticleArea.music:
          return MusicArticle();
      }
      return Article();
    };
```

Here's a complete example with nested objects and complex list elements:
```dart
enum ArticleArea { electronics, music }

class Article extends SerializableObject {
  Article() {
    transformers['area'] = (value) =>
        value is ArticleArea ? value.index : ArticleArea.values[value];
  }

  ArticleArea get area => attributes['area'];
  set area(ArticleArea value) => attributes['area'] = value;

  String get name => attributes['name'];
  set name(String value) => attributes['name'] = value;

  int get price => attributes['price'];
  set price(int value) => attributes['price'] = value;
}

class ElectronicsArticle extends Article {
  ElectronicsArticle() {
    area = ArticleArea.electronics;
  }

  String get recommendation => attributes['recommendation'];
  set recommendation(String value) => attributes['recommendation'] = value;
}

class MusicArticle extends Article {
  MusicArticle() {
    area = ArticleArea.music;
    objectCreators['band'] = (map) => Band();
  }

  Band get band => attributes['band'];
  set band(Band value) => attributes['band'] = value;
}

class Band extends SerializableObject {
  String get name => attributes['name'];
  set name(String value) => attributes['name'] = value;

  int get year => attributes['year'];
  set year(int value) => attributes['year'] = value;

  Band({String name, int year}) {
    this.name = name;
    this.year = year;
  }
}

class Order extends SerializableObject {
  Order() {
    objectCreators['articles'] = (map) => <Article>[];
    objectCreators['articles.value'] = (map) {
      final int areaIndex = map['area'];
      final area = ArticleArea.values[areaIndex];
      switch (area) {
        case ArticleArea.electronics:
          return ElectronicsArticle();
        case ArticleArea.music:
          return MusicArticle();
      }
      return Article();
    };
  }

  List<Article> get articles => attributes['articles'];
  set articles(List<Article> value) => attributes['articles'] = value;
}

void main() {
  final order = Order()
    ..articles = [
      ElectronicsArticle()
        ..name = 'CD Player'
        ..price = 3799
        ..recommendation = 'Consider our streaming option, too!',
      ElectronicsArticle()
        ..name = 'MC Tape Deck'
        ..price = 12399
        ..recommendation = 'Old school, like it!',
      MusicArticle()
        ..name = 'The white album'
        ..price = 1899
        ..band = Band(name: 'Beatles', year: 1962)
    ];
  final serializer = Serializer();
  final json = serializer.serialize(order);
  print('order: $json');

  final inputJson =
      '{"articles": [{"area": 0, "name": "CD Player", "price": 3799, "recommendation": "Consider our streaming option, too!"}, '
      '{"area": 0, "name": "MC Tape Deck", "price": 12399, "recommendation": "Old school, like it!"}, '
      '{"area": 1, "name": "The white album", "price": 1899, "band": {"name": "Beatles", "year": 1962}}]}';
  final deserializedOrder = Order();
  serializer.deserialize(inputJson, deserializedOrder);
  for (var i = 0; i < deserializedOrder.articles.length; i++) {
    final article = deserializedOrder.articles[i];
    print('$i: area: ${article.area}');
    print('$i: name: ${article.name}');
    print('$i: price: ${article.price}');
    if (article is ElectronicsArticle) {
      print('$i: recommendation: ${article.recommendation}');
    } else if (article is MusicArticle) {
      print('$i: band-name: ${article.band.name}');
      print('$i: band-year: ${article.band.year}');
    }
  }
}
```

#### Nested Maps
When nesting maps, you also register your creator function in the `objectCreators`, e.g. 
```dart
objectCreators['news-by-year'] = (map) => <int, String>{};
```
When dealing with maps with non-String keys, you need to transform these to Strings and back to their original 
form when deserializing. As for `enum`s you do this by registering a `transformer` function that receives the value. 
You registers the function under the field name with a `.key` appended to it:
```dart
    transformers['news-by-year.key'] =
        (value) => value is int ? value.toString() : int.parse(value);
```

Same as for lists, you can also register objectCreators and transformers for map values by appending a `.value` to the field's serialization name. 

Here is a complete example for serializing an object with an embedded map:

```dart
class MappedArticle extends SerializableObject {
  MappedArticle() {
    objectCreators['news-by-year'] = (map) => <int, String>{};
    transformers['news-by-year.key'] =
        (value) => value is int ? value.toString() : int.parse(value);
  }

  String get name => attributes['name'];
  set name(String value) => attributes['name'] = value;

  Map<int, String> get newsByYear => attributes['news-by-year'];
  set newsByYear(Map<int, String> value) => attributes['news-by-year'] = value;
}

void main() {
  final newsByYear = {
    2020: 'Corona, Corona, Corona...',
    2021: 'The end of a pandemia',
    2022: 'Climate change getting really serious'
  };
  final article = MappedArticle()
    ..name = 'My Article'
    ..newsByYear = newsByYear;
  final serializer = Serializer();
  final json = serializer.serialize(article);
  print('article with map: $json');

  final inputJson =
      '{"name": "My Article", "news-by-year": {"2020": "Corona, Corona, Corona...", "2021": "The end of a pandemic", "2022": "Climate change getting really serious"}}';
  final deserializedArticle = MappedArticle();
  serializer.deserialize(inputJson, deserializedArticle);
  print('article: ${article.name}');
  for (final key in article.newsByYear.keys) {
    print('$key: ${article.newsByYear[key]}');
  }
}
```

### On Demand Serialization
Using a dynamic map for storing and retrieving fields requires you to create boiler-plate code and makes accessing fields slower.
When you want to use normal fields, you can implement `OnDemandSerializable` instead:
```dart
class Article implements OnDemandSerializable {
  int price;
  String name;
  

  @override
  void write(Map<String, dynamic> attributes) {
    attributes['price'] = price;
    attributes['name'] = name;
  }

  @override
  void read(Map<String, dynamic> attributes) {
    price = attributes['price'];
    name = attributes['name'];
  }
```

Use `Serializer.serializeOnDemand(OnDemandSerializable)` to create the JSON and `Serializer.deserializeOnDemand(String,OnDemandSerializable)`
to deserialize your classes.

In these calls you can optionally provide `transformers` and `objectCreators` to handle more complex scenarios like `enum`s, other nested `OnDemandSerializable`, `List` or `Map` fields. The registration is the same as for `Serializable`, however you have can only register these special cases at that point. This means that field names of special cases must be unique throughout your object's hierarchy.

Here's a complete example of a complex structure with `OnDemandSerializable`:
```dart
class OnDemandArticle implements OnDemandSerializable {
  String name;
  Map<int, String> newsByYear;

  String serialize() {
    final serializer = Serializer();
    final json = serializer.serializeOnDemand(
      this,
      transformers: {
        'news-by-year.key': (value) =>
            value is int ? value.toString() : int.parse(value),
      },
    );
    return json;
  }

  void deserialize(String json) {
    final serializer = Serializer();
    serializer.deserializeOnDemand(
      json,
      this,
      transformers: {
        'news-by-year.key': (value) =>
            value is int ? value.toString() : int.parse(value),
      },
      objectCreators: {
        'news-by-year': (map) => <int, String>{},
      },
    );
  }

  @override
  void write(Map<String, dynamic> attributes) {
    attributes['name'] = name;
    attributes['news-by-year'] = newsByYear;
  }

  @override
  void read(Map<String, dynamic> attributes) {
    name = attributes['name'];
    newsByYear = attributes['news-by-year'];
  }
}

void main() {
  final newsByYear = {
    2020: 'Corona, Corona, Corona...',
    2021: 'The end of a pandemia',
    2022: 'Climate change getting really serious'
  };
  final article = OnDemandArticle()
    ..name = 'My Article'
    ..newsByYear = newsByYear;
  final json = article.serialize();
  print('on demand article: $json');

  final inputJson =
      '{"name": "My Article", "news-by-year": {"2020": "Corona, Corona, Corona...", "2021": "The end of a pandemic", "2022": "Climate change getting really serious"}}';
  final deserializedArticle = OnDemandArticle();
  deserializedArticle.deserialize(inputJson);
  print('deserialized article: ${article.name}');
  for (final key in article.newsByYear.keys) {
    print('$key: ${article.newsByYear[key]}');
  }
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/Enough-Software/enough_serialization/issues

## Null-Safety

enough_convert is null-safe from v1.3.0 onward.

## License

Licensed under the commercial friendly [MIT License](LICENSE).

CHANGELOG.md
## 1.4.0

- Fixed deserialization bug when String value contains a newline character, #5 

## 1.3.0

- Add null-safety.

## 1.2.0

- Add option to directly serialize/deserialize list values with `Serializer.serializeList(...)` and `Serializer.deserializeList(...)`.

## 1.1.0

- Support `Map` fields
- Support on demand serilization by extending `OnDemandSerializable`or `OndDemandSerializableObject`.

## 1.0.1

- Improve documentation

## 1.0.0

- Initial version


test/src/serializable_test.dart
import 'dart:convert';

import 'package:enough_serialization/src/serializable.dart';
import 'package:test/test.dart';

void main() {
  group('Serilization Tests', () {
    late MySerializable mySerializable;

    setUp(() {
      final props = {'firstname': 'firstValue', 'second': 'another value'};
      final newsByYear = {
        2020: 'enough serialization started',
        2021: 'the future is there'
      };
      final child = MySerializable()
        ..myText = 'hi this is the child element'
        ..myNumber = 23
        ..myChild = null;
      mySerializable = MySerializable()
        ..myNumber = 13
        ..myEnum = MySerializableEnum.value3
        ..myText = 'Hello "World"'
        ..myList = ['one', 'two', 'three']
        ..myChild = child
        ..events = [
          StartEvent()..players = [232, 12, 2423, 99],
          EndEvent()..winner = 2423
        ]
        ..properties = props
        ..newsByYear = newsByYear;
    });

    test('Serialize', () {
      final serializer = Serializer();
      final json = serializer.serialize(mySerializable);
      print(json);
      expect(json, isNotNull);
      final decoded = JsonCodec().decode(json);
      expect(decoded, isNotNull);
      print(decoded);
    });

    test('Deserialize', () {
      final jsonText =
          r'{"my-number": 13, "my-enum": 2, "my-text": "Hello \"World\"", "my-list": ["one", "two", "three"], "my-serializable": {"my-text": "hi this is the child element", "my-number": 23, "my-serializable": null}, "events": [{"type": 0, "players": [232, 12, 2423, 99]}, {"type": 1, "winner": 2423}], "properties": {"firstname": "firstValue", "second": "another value"}, "news-by-year": {"2020": "enough serialization started", "2021": "the future is there"}}';

      final serializer = Serializer();
      final serializable = MySerializable();
      serializer.deserialize(jsonText, serializable);
      expect(serializable.myNumber, 13);
      expect(serializable.myEnum, MySerializableEnum.value3);
      expect(serializable.myText, 'Hello "World"');
      expect(serializable.myList, isNotEmpty);
      expect(serializable.myList!.length, 3);
      expect(serializable.myList, ['one', 'two', 'three']);
      expect(serializable.myChild, isNotNull);
      expect(serializable.myChild!.myText, 'hi this is the child element');
      expect(serializable.myChild!.myNumber, 23);
      expect(serializable.events, isNotEmpty);
      expect(serializable.events!.length, 2);
      expect(serializable.events![0], isA<StartEvent>());
      expect(serializable.events![0].type, EventType.start);
      expect(
          (serializable.events![0] as StartEvent).players, [232, 12, 2423, 99]);

      expect(serializable.events![1], isA<EndEvent>());
      expect(serializable.events![1].type, EventType.end);
      expect((serializable.events![1] as EndEvent).winner, 2423);
      expect(serializable.properties, isNotNull);
      expect(serializable.properties!['firstname'], 'firstValue');
      expect(serializable.properties!['second'], 'another value');
      expect(serializable.newsByYear![2020], 'enough serialization started');
      expect(serializable.newsByYear![2021], 'the future is there');
    });
  });
}

enum MySerializableEnum { value1, value2, value3 }

enum EventType { start, end }

class Event extends SerializableObject {
  EventType? get type => attributes['type'];
  set type(EventType? value) => attributes['type'] = value;

  Event(EventType type) {
    this.type = type;
    transformers['type'] =
        (value) => value is EventType ? value.index : EventType.values[value];
  }
}

class StartEvent extends Event {
  StartEvent() : super(EventType.start) {
    objectCreators['players'] = (map) => <int>[];
  }

  List<int>? get players => attributes['players'];
  set players(List<int>? value) => attributes['players'] = value;
}

class EndEvent extends Event {
  EndEvent() : super(EventType.end);

  int? get winner => attributes['winner'];
  set winner(int? value) => attributes['winner'] = value;
}

class MySerializable extends SerializableObject {
  MySerializable() {
    objectCreators['my-list'] = (map) => <String>[];
    objectCreators['events'] = (map) => <Event>[];
    objectCreators['my-serializable'] = (map) => MySerializable();
    objectCreators['events.value'] =
        (map) => map!['type'] == 0 ? StartEvent() : EndEvent();
    objectCreators['properties'] = (map) => <String, String>{};
    objectCreators['news-by-year'] = (map) => <int, String>{};
    transformers['news-by-year.key'] =
        (value) => value is int ? value.toString() : int.parse(value);
    transformers['my-enum'] = (value) => value is MySerializableEnum
        ? value.index
        : MySerializableEnum.values[value];
  }

  List<Event>? get events => attributes['events'];
  set events(List<Event>? value) => attributes['events'] = value;

  MySerializableEnum? get myEnum => attributes['my-enum'];
  set myEnum(MySerializableEnum? value) => attributes['my-enum'] = value;

  int? get myNumber => attributes['my-number'];
  set myNumber(int? value) => attributes['my-number'] = value;

  String? get myText => attributes['my-text'];
  set myText(String? value) => attributes['my-text'] = value;

  List<String>? get myList => attributes['my-list'];
  set myList(List<String>? value) => attributes['my-list'] = value;

  MySerializable? get myChild => attributes['my-serializable'];
  set myChild(MySerializable? value) => attributes['my-serializable'] = value;

  Map<String, String>? get properties => attributes['properties'];
  set properties(Map<String, String>? value) => attributes['properties'] = value;

  Map<int, String>? get newsByYear => attributes['news-by-year'];
  set newsByYear(Map<int, String>? value) => attributes['news-by-year'] = value;
}


test/src/ondemandserializable_test.dart
import 'dart:convert';

import 'package:enough_serialization/src/serializable.dart';
import 'package:test/test.dart';

void main() {
  group('On Demand Serilization Tests', () {
    late MySerializable mySerializable;

    setUp(() {
      final props = {'firstname': 'firstValue', 'second': 'another value'};
      final newsByYear = {
        2020: 'enough serialization started',
        2021: 'the future is there'
      };
      final child = MySerializable()
        ..myText = 'hi this is the child element'
        ..myNumber = 23
        ..myChild = null;
      mySerializable = MySerializable()
        ..myNumber = 13
        ..myEnum = MySerializableEnum.value3
        ..myText = 'Hello "World"'
        ..myList = ['one', 'two', 'three']
        ..myChild = child
        ..events = [
          StartEvent()..players = [232, 12, 2423, 99],
          EndEvent()..winner = 2423
        ]
        ..properties = props
        ..newsByYear = newsByYear;
    });

    test('Serialize on demand', () {
      final json = mySerializable.serialize();
      print(json);
      expect(json, isNotNull);
      final decoded = JsonCodec().decode(json);
      expect(decoded, isNotNull);
      print(decoded);
    });

    test('Deserialize on demand', () {
      final jsonText =
          r'{"my-number": 13, "my-enum": 2, "my-text": "Hello \"World\"", "my-list": ["one", "two", "three"], "my-serializable": {"my-text": "hi this is the child element", "my-number": 23, "my-serializable": null}, "events": [{"type": 0, "players": [232, 12, 2423, 99]}, {"type": 1, "winner": 2423}], "properties": {"firstname": "firstValue", "second": "another value"}, "news-by-year": {"2020": "enough serialization started", "2021": "the future is there"}}';

      final serializable = MySerializable.deserialize(jsonText);
      expect(serializable.myNumber, 13);
      expect(serializable.myEnum, MySerializableEnum.value3);
      expect(serializable.myText, 'Hello "World"');
      expect(serializable.myList, isNotEmpty);
      expect(serializable.myList!.length, 3);
      expect(serializable.myList, ['one', 'two', 'three']);
      expect(serializable.myChild, isNotNull);
      expect(serializable.myChild!.myText, 'hi this is the child element');
      expect(serializable.myChild!.myNumber, 23);
      expect(serializable.events, isNotEmpty);
      expect(serializable.events!.length, 2);
      expect(serializable.events![0], isA<StartEvent>());
      expect(serializable.events![0].type, EventType.start);
      expect(
          (serializable.events![0] as StartEvent).players, [232, 12, 2423, 99]);

      expect(serializable.events![1], isA<EndEvent>());
      expect(serializable.events![1].type, EventType.end);
      expect((serializable.events![1] as EndEvent).winner, 2423);
      expect(serializable.properties, isNotNull);
      expect(serializable.properties!['firstname'], 'firstValue');
      expect(serializable.properties!['second'], 'another value');
      expect(serializable.newsByYear![2020], 'enough serialization started');
      expect(serializable.newsByYear![2021], 'the future is there');
    });
  });
}

enum MySerializableEnum { value1, value2, value3 }

enum EventType { start, end }

class Event implements OnDemandSerializable {
  EventType? type;

  Event(this.type);

  @override
  void read(Map<String, dynamic> attributes) {
    type = attributes['type'];
  }

  @override
  void write(Map<String, dynamic> attributes) {
    attributes['type'] = type;
  }
}

class StartEvent extends Event {
  StartEvent() : super(EventType.start);

  List<int>? players;

  @override
  void read(Map<String, dynamic> attributes) {
    super.read(attributes);
    players = attributes['players'];
  }

  @override
  void write(Map<String, dynamic> attributes) {
    super.write(attributes);
    attributes['players'] = players;
  }
}

class EndEvent extends Event {
  EndEvent() : super(EventType.end);

  int? winner;

  @override
  void read(Map<String, dynamic> attributes) {
    super.read(attributes);
    winner = attributes['winner'];
  }

  @override
  void write(Map<String, dynamic> attributes) {
    super.write(attributes);
    attributes['winner'] = winner;
  }
}

class MySerializable implements OnDemandSerializable {
  List<Event>? events;
  MySerializableEnum? myEnum;
  int? myNumber;
  String? myText;
  List<String>? myList;
  MySerializable? myChild;
  Map<String, String>? properties;
  Map<int, String>? newsByYear;

  String serialize() {
    final serializer = Serializer();
    return serializer.serializeOnDemand(
      this,
      transformers: {
        'news-by-year.key': (value) =>
            value is int ? value.toString() : int.parse(value),
        'my-enum': (value) => value is MySerializableEnum
            ? value.index
            : MySerializableEnum.values[value],
        // for event classes:
        'type': (value) =>
            value is EventType ? value.index : EventType.values[value],
      },
    );
  }

  static MySerializable deserialize(String jsonText) {
    final serializer = Serializer();
    final target = MySerializable();
    serializer.deserializeOnDemand(
      jsonText,
      target,
      transformers: {
        'news-by-year.key': (value) =>
            value is int ? value.toString() : int.parse(value),
        'my-enum': (value) => value is MySerializableEnum
            ? value.index
            : MySerializableEnum.values[value],
        // for event classes:
        'type': (value) =>
            value is EventType ? value.index : EventType.values[value],
      },
      objectCreators: {
        'my-list': (map) => <String>[],
        'events': (map) => <Event>[],
        'my-serializable': (map) => MySerializable(),
        'events.value': (map) => map!['type'] == 0 ? StartEvent() : EndEvent(),
        'properties': (map) => <String, String>{},
        'news-by-year': (map) => <int, String>{},
        // for StartEvent instances:
        'players': (map) => <int>[],
      },
    );
    return target;
  }

  @override
  void read(Map<String, dynamic> attributes) {
    events = attributes['events'];
    myEnum = attributes['my-enum'];
    myNumber = attributes['my-number'];
    myText = attributes['my-text'];
    myList = attributes['my-list'];
    myChild = attributes['my-serializable'];
    properties = attributes['properties'];
    newsByYear = attributes['news-by-year'];
  }

  @override
  void write(Map<String, dynamic> attributes) {
    attributes['events'] = events;
    attributes['my-enum'] = myEnum;
    attributes['my-number'] = myNumber;
    attributes['my-text'] = myText;
    attributes['my-list'] = myList;
    attributes['my-serializable'] = myChild;
    attributes['properties'] = properties;
    attributes['news-by-year'] = newsByYear;
  }
}


example/enough_serialization_example.dart
import 'package:enough_serialization/enough_serialization.dart';

void main() {
  simpleExample();
  print('');
  enumExample();
  print('');
  complexExample();
  print('');
  mapExample();
  print('');
  onDemandExample();
}

class SimpleArticle extends SerializableObject {
  String? get name => attributes['name'];
  set name(String? value) => attributes['name'] = value;

  int? get price => attributes['price'];
  set price(int? value) => attributes['price'] = value;

  double? get popularity => attributes['popularity'];
  set popularity(double? value) => attributes['popularity'] = value;
}

void simpleExample() {
  print('simple example:');
  final article = SimpleArticle()
    ..name = 'Remote Control'
    ..price = 2499
    ..popularity = 0.1;
  final serializer = Serializer();
  final json = serializer.serialize(article);
  print('serialized article: $json');

  final inputJson =
      '{"name": "Remote Control", "price": 2499, "popularity": 0.1}';
  final deserializedArticle = SimpleArticle();
  serializer.deserialize(inputJson, deserializedArticle);
  print('name: ${deserializedArticle.name}');
  print('price: ${deserializedArticle.price}');
  print('popularity: ${deserializedArticle.popularity}');
}

enum ArticleArea { electronics, music }

class ArticleWithEnum extends SerializableObject {
  ArticleWithEnum() {
    transformers['area'] = (value) =>
        value is ArticleArea ? value.index : ArticleArea.values[value];
  }

  ArticleArea? get area => attributes['area'];
  set area(ArticleArea? value) => attributes['area'] = value;

  String? get name => attributes['name'];
  set name(String? value) => attributes['name'] = value;
}

void enumExample() {
  print('emum example:');
  final article = ArticleWithEnum()
    ..area = ArticleArea.electronics
    ..name = 'Remote Control';
  final serializer = Serializer();
  final json = serializer.serialize(article);
  print('serialized article: $json');

  final inputJson = '{"area": 0, "name": "Remote Control"}';
  final deserializedArticle = ArticleWithEnum();
  serializer.deserialize(inputJson, deserializedArticle);
  print('area: ${deserializedArticle.area}');
  print('name: ${deserializedArticle.name}');
}

class Article extends SerializableObject {
  Article() {
    transformers['area'] = (value) =>
        value is ArticleArea ? value.index : ArticleArea.values[value];
  }

  ArticleArea? get area => attributes['area'];
  set area(ArticleArea? value) => attributes['area'] = value;

  String? get name => attributes['name'];
  set name(String? value) => attributes['name'] = value;

  int? get price => attributes['price'];
  set price(int? value) => attributes['price'] = value;
}

class ElectronicsArticle extends Article {
  ElectronicsArticle() {
    area = ArticleArea.electronics;
  }

  String? get recommendation => attributes['recommendation'];
  set recommendation(String? value) => attributes['recommendation'] = value;
}

class MusicArticle extends Article {
  MusicArticle() {
    area = ArticleArea.music;
    objectCreators['band'] = (map) => Band();
  }

  Band? get band => attributes['band'];
  set band(Band? value) => attributes['band'] = value;
}

class Band extends SerializableObject {
  String? get name => attributes['name'];
  set name(String? value) => attributes['name'] = value;

  int? get year => attributes['year'];
  set year(int? value) => attributes['year'] = value;

  Band({String? name, int? year}) {
    this.name = name;
    this.year = year;
  }
}

class Order extends SerializableObject {
  Order() {
    objectCreators['articles'] = (map) => <Article>[];
    // the list articles contains different Articel instances depending on the specified area field:
    objectCreators['articles.value'] = (map) {
      final int areaIndex = map!['area'];
      final area = ArticleArea.values[areaIndex];
      switch (area) {
        case ArticleArea.electronics:
          return ElectronicsArticle();
        case ArticleArea.music:
          return MusicArticle();
      }
      return Article();
    };
  }

  List<Article>? get articles => attributes['articles'];
  set articles(List<Article>? value) => attributes['articles'] = value;
}

void complexExample() {
  print('complex example:');
  final order = Order()
    ..articles = [
      ElectronicsArticle()
        ..name = 'CD Player'
        ..price = 3799
        ..recommendation = 'Consider our streaming option, too!',
      ElectronicsArticle()
        ..name = 'MC Tape Deck'
        ..price = 12399
        ..recommendation = 'Old school, like it!',
      MusicArticle()
        ..name = 'The white album'
        ..price = 1899
        ..band = Band(name: 'Beatles', year: 1962)
    ];
  final serializer = Serializer();
  final json = serializer.serialize(order);
  print('order: $json');

  final inputJson =
      '{"articles": [{"area": 0, "name": "CD Player", "price": 3799, "recommendation": "Consider our streaming option, too!"}, '
      '{"area": 0, "name": "MC Tape Deck", "price": 12399, "recommendation": "Old school, like it!"}, '
      '{"area": 1, "name": "The white album", "price": 1899, "band": {"name": "Beatles", "year": 1962}}]}';
  final deserializedOrder = Order();
  serializer.deserialize(inputJson, deserializedOrder);
  for (var i = 0; i < deserializedOrder.articles!.length; i++) {
    final article = deserializedOrder.articles![i];
    print('$i: area: ${article.area}');
    print('$i: name: ${article.name}');
    print('$i: price: ${article.price}');
    if (article is ElectronicsArticle) {
      print('$i: recommendation: ${article.recommendation}');
    } else if (article is MusicArticle) {
      print('$i: band-name: ${article.band!.name}');
      print('$i: band-year: ${article.band!.year}');
    }
  }
}

class MappedArticle extends SerializableObject {
  MappedArticle() {
    objectCreators['news-by-year'] = (map) => <int, String>{};
    transformers['news-by-year.key'] =
        (value) => value is int ? value.toString() : int.parse(value);
  }

  String? get name => attributes['name'];
  set name(String? value) => attributes['name'] = value;

  Map<int, String>? get newsByYear => attributes['news-by-year'];
  set newsByYear(Map<int, String>? value) => attributes['news-by-year'] = value;
}

void mapExample() {
  print('map example:');
  final newsByYear = {
    2020: 'Corona, Corona, Corona...',
    2021: 'The end of a pandemia',
    2022: 'Climate change getting really serious'
  };
  final article = MappedArticle()
    ..name = 'My Article'
    ..newsByYear = newsByYear;
  final serializer = Serializer();
  final json = serializer.serialize(article);
  print('article with map: $json');

  final inputJson =
      '{"name": "My Article", "news-by-year": {"2020": "Corona, Corona, Corona...", "2021": "The end of a pandemic", "2022": "Climate change getting really serious"}}';
  final deserializedArticle = MappedArticle();
  serializer.deserialize(inputJson, deserializedArticle);
  print('deserialized article: ${article.name}');
  for (final key in article.newsByYear!.keys) {
    print('$key: ${article.newsByYear![key]}');
  }
}

class OnDemandArticle implements OnDemandSerializable {
  String? name;
  Map<int, String>? newsByYear;

  String serialize() {
    final serializer = Serializer();
    final json = serializer.serializeOnDemand(
      this,
      transformers: {
        'news-by-year.key': (value) =>
            value is int ? value.toString() : int.parse(value),
      },
    );
    return json;
  }

  void deserialize(String json) {
    final serializer = Serializer();
    serializer.deserializeOnDemand(
      json,
      this,
      transformers: {
        'news-by-year.key': (value) =>
            value is int ? value.toString() : int.parse(value),
      },
      objectCreators: {
        'news-by-year': (map) => <int, String>{},
      },
    );
  }

  @override
  void write(Map<String, dynamic> attributes) {
    attributes['name'] = name;
    attributes['news-by-year'] = newsByYear;
  }

  @override
  void read(Map<String, dynamic> attributes) {
    name = attributes['name'];
    newsByYear = attributes['news-by-year'];
  }
}

void onDemandExample() {
  print('on demand example:');
  final newsByYear = {
    2020: 'Corona, Corona, Corona...',
    2021: 'The end of a pandemia',
    2022: 'Climate change getting really serious'
  };
  final article = OnDemandArticle()
    ..name = 'My Article'
    ..newsByYear = newsByYear;
  final json = article.serialize();
  print('on demand article: $json');

  final inputJson =
      '{"name": "My Article", "news-by-year": {"2020": "Corona, Corona, Corona...", "2021": "The end of a pandemic", "2022": "Climate change getting really serious"}}';
  final deserializedArticle = OnDemandArticle();
  deserializedArticle.deserialize(inputJson);
  print('deserialized article: ${article.name}');
  for (final key in article.newsByYear!.keys) {
    print('$key: ${article.newsByYear![key]}');
  }
}
